### PR TITLE
Add support for 'X-Forwarded-For' header

### PR DESCRIPTION
**Pull request by vbelov42**
Usually people care about privacy and anonymity on the Internet, so proxy software is frequently used to filter out or mange several headers, to hide client's identity. But sometimes situation is completely different and one needs to tell the server about client's real address. Here comes a 'X-Forwarded-For' header that was introduced by the Squid caching proxy server's developers.

> The X-Forwarded-For (XFF) HTTP header field is a common method for identifying the originating IP address of a client connecting to a web server through an HTTP proxy or load balancer. [[wikipedia](https://en.wikipedia.org/wiki/X-Forwarded-For)]

'X-Tinyproxy' field serves the same target, but is much less recognized.

In a nutshell, I just mutated ```write_via_header()``` function into ```write_xff_header()```. And placed a call for it after ```write_via_header()``` in processing. I'm not 100% positive that it's correct for ```process_server_headers()``` case.

Also I added an option to the configuration file to control this. It has some description and default value that disables this feature.

### CODE REVIEW

**Comment by rofl0r** [comment]
> instead of leaving the default behaviour as is and giving an option to enable it

i suppose you meant *disable* here ?

**Comment by vbelov42** [comment]
I followed the path of 'Via' header, that serves the same task. It is controlled by 'disable' switch config.disable_viaheader. I was surprised by the way how it is done, as you were, but just copied the existing behaviour. More surprising is that there is no default setting for this, so without configuration file it'll be enabled.
I see two ways: rename both options, or set a 'disable' default value. Remember, that changing a variable name for the option will call for a change of a field name in configuration file, and that requires updating of all configurations. So right now I'd suggest the latter way.

**Comment by vbelov42** [comment]
Also I just realized that I forgot to add a proper line to ```initialize_with_defaults(...)``` for ```disable_xffheader```.

**Comment by obnoxxx** [comment]
@rofl0r wrote:
> >    instead of leaving the default behaviour as is and giving an option to enable it
> 
> i suppose you meant disable here ?

No. The current patch enables the new header by default and offers a disable option. I was suggesting to do the reverse.

**Comment by obnoxxx** [comment]
@vbelov42 wrote:
> I followed the path of 'Via' header, that serves similar task. It is controlled by 'disable' switch `config.disable_viaheader`.  I was surprised by the way how it is done, as you were, but just copied  the existing behaviour. More surprising is that there is no default  setting for this, so without configuration file it'll be enabled.  Manpage indicates that this might be intentional, but I don't completely  follow that.
> 
> I see two ways: rename both options, or set a 'disable' default  value. Remember, that changing a variable name for the option will call  for a change of a field name in configuration file, and that requires  updating of all configurations. So right now I'd suggest the latter way.

Tinyproxy currently has no capability of adding the XFF header. So the default behaviour is not to set it. The reasonable way to introduce the new feature (according to me, quite generally) should be to not change the current behaviour but leave it as is and maybe change the default in a later release if desired. I.e. you would have to actively do a configuration change in order to trigger the new behaviour (setting the XFF header).

What adds to this, imho, is that the XFF header may not be universally desired, since it exposes the client IP to the server on the other side of the proxy. So if we take the behaviour proposed in the current patch, a user would suddenly get "unstealth-ed" by an upgrade.

The fact that the via header is enabled by default is not relevant: The via header is an existing capability. This here is about introducing a new capability. It is more about expected behaviour than about similarity of code.

**Comment by vbelov42** [comment]
Just for the case, I did what you requested. It's in 5df3615.

**Comment by captn3m0** [comment]
Can this be merged?

(Aside: The via header is a requirement as per the RFC, which is why it needs to be enabled by default)

**Comment by rofl0r** [comment]
"the RFC" ? please be more specific.

**Comment by captn3m0** [comment]
[RFC 7230](https://tools.ietf.org/html/rfc7230#section-5.7.1) marks it as a `MUST`. [CloudFlare](https://blog.cloudflare.com/preventing-malicious-request-loops/) has a nice blog post about why it is necessary.

**Comment by rofl0r** [comment]
both the RFC and the cloudflare blog talk about the `Via` header (which is already supported), not the `XFF` header...

**Comment by captn3m0** [comment]
Yeah, mentioned that in my comment. Was just giving reason why the Via header is default On. See above for some discussion around it.

**Comment by rofl0r** [comment]
oh, you're right. :+1:

**Comment by juju4** [comment]
Is this PR still under consideration?
XFF header is good in company context and/or chaining proxies.

**Comment by dk-jsco** [comment]
TinyProxy is a great project - but it would be even better with XFF. 

Especially when working with backend services running in k8s. 
The **X-Tinyproxy** header is removed by Ingress - hence, it's not possible to preserve the client IP.
_NOTE: This is not the case with **X-Forwarded-For**_

It is perfectly fine to disable XFF by default

Therefore, please provide a status on this pull request

**Comment by rofl0r** [comment]
> The X-Tinyproxy header is removed by Ingress - hence, it's not possible to preserve the client IP.

that sounds like an ingress problem then. maybe you should take it there.

> Therefore, please provide a status on this pull request

it would need some work for infrastructure changes that happened related to conf parsing. as the internet gets more and more hostile against proxy users (thanks cloudflare) i personally am not eager to add yet another option that removes privacy.
